File: Python/Stage 3/S3_PIPE_06.py
def pipe_safe(*fns):
    def apply(value):
        result = value
        try:
            for fn in fns:
                result = fn(result)
            return {"ok": True, "value": result}
        except Exception as exc:
            return {"ok": False, "error": str(exc)}

    return apply


def inverse(n):
    if n == 0:
        raise ValueError("division by zero")
    return 1 / n


safe = pipe_safe(lambda x: x + 1, inverse)
print(safe(1))
print(safe(-1))

File: Js/Task 6/S3_PIPE_06.js
function pipeSafe(...fns) {
    return x => {
        try {
            let result = x;
            for (let fn of fns) {
                result = fn(result);
            }
            return { ok: true, value: result };
        } catch (error) {
            return { ok: false, error: error.message };
        }
    };
}

function add(x) {
    return x + 1;
}

function divide(x) {
    if (x === 0) throw new Error("Cannot divide by zero");
    return x / 2;
}

const safeProcess = pipeSafe(add, divide);

console.log(safeProcess(4));  // { ok: true, value: 2.5 }
console.log(safeProcess(0));  // { ok: false, error: "Cannot divide by zero" }